### DISCUSSION TITLE
How to walk routing

### DISCUSSION THREAD

**Discussion by JonathanCabezas** [RobinTail/express-zod-api#2257] [Q&A] [answered]
Hello,

I've tried to walk the routing object, as you do to generate the documentation, but `walkRouting` is not exposed, `AbstractEndpoint` neither. It makes it pretty hard to reproduce this. How would you go about it ?

My 2 usages are:

1. To write a test to verify that my documentation has the right number of expected routes
2. I want to generate a ts file to have autocomplete on my routes

For usage 2, do you see another way to do this ?

I already tried with static type analysis, using [NestedKeyOf](https://dev.to/pffigueiredo/typescript-utility-keyof-nested-object-2pa3) but my ts server was getting awfully slow for autocomplete.

**Comment by JonathanCabezas** [comment]
Here's what I came up with to count routes:
```typescript
import { DependsOnMethod, Routing } from "express-zod-api";

const countNbEndpoints = (obj: Routing) => {
  let nbEndpoints = 0;
  const keys = Object.keys(obj);

  // ENDPOINT LEAF
  if (keys.length === 0) {
    return 1;
  }

  keys.forEach(key => {
    const value = obj[key];

    // DEPENDS ON METHOD LEAF
    if (value instanceof DependsOnMethod) {
      nbEndpoints += value.entries.length;
      return;
    }

    // NODE
    if (typeof value === "object" && value !== null) {
      nbEndpoints += countNbEndpoints(value as Routing);
    }
  });

  return nbEndpoints;
};
```

I'm not using `StaticServe` for now.

**Comment by RobinTail** [comment, marked as answer]
> To write a test to verify that my documentation has the right number of expected routes
> I want to generate a ts file to have autocomplete on my routes

How about this, @JonathanCabezas ?

```ts
const documentation = new Documentation({
  routing,
  config,
  version: "1.2.3",
  title: "Example API",
  serverUrl: "https://example.com",
});

console.log(
  Object.keys(documentation.rootDoc.paths!)
);
```

![image](https://github.com/user-attachments/assets/3775bfad-384a-4c99-a5eb-9d34a5c4a456)
